File: extension/src/main/java/io/holunda/camunda/bpm/data/adapter/set/SetReadWriteAdapterVariableScope.java
package io.holunda.camunda.bpm.data.adapter.set;

import org.camunda.bpm.engine.delegate.VariableScope;

import java.util.Optional;
import java.util.Set;

/**
 * Read-write adapter for variable scope.
 *
 * @param <T> type of value.
 */
public class SetReadWriteAdapterVariableScope<T> extends AbstractSetReadWriteAdapter<T> {

    private final VariableScope variableScope;

    /**
     * Constructs the adapter.
     *
     * @param variableScope variable scope to access.
     * @param variableName  variable to access.
     * @param memberClazz   class of member variable value.
     */
    public SetReadWriteAdapterVariableScope(VariableScope variableScope, String variableName, Class<T> memberClazz) {
        super(variableName, memberClazz);
        this.variableScope = variableScope;
    }

    @Override
    public Optional<Set<T>> getOptional() {
        return Optional.ofNullable(getOrNull(variableScope.getVariable(variableName)));
    }

    @Override
    public void set(Set<T> value, boolean isTransient) {
        variableScope.setVariable(variableName, getTypedValue(value, isTransient));
    }

    @Override
    public Optional<Set<T>> getLocalOptional() {
        return Optional.ofNullable(getOrNull(variableScope.getVariableLocal(variableName)));
    }

    @Override
    public void setLocal(Set<T> value, boolean isTransient) {
        variableScope.setVariableLocal(variableName, getTypedValue(value, isTransient));
    }

    @Override
    public void remove() {
        variableScope.removeVariable(variableName);
    }

    @Override
    public void removeLocal() {
        variableScope.removeVariableLocal(variableName);
    }

}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/adapter/set/SetReadWriteAdapterVariableScope.kt
package io.holunda.camunda.bpm.data.adapter.set

import org.camunda.bpm.engine.delegate.VariableScope
import java.util.*

/**
 * Read-write adapter for variable scope.
 *
 * @param [T] type of value.
 * @param variableScope variable scope to access.
 * @param variableName  variable to access.
 * @param memberClazz   class of member variable value.
 */
class SetReadWriteAdapterVariableScope<T>(
  private val variableScope: VariableScope,
  variableName: String,
  memberClazz: Class<T>
) : AbstractSetReadWriteAdapter<T>(variableName, memberClazz) {
  override fun getOptional(): Optional<Set<T>> {
    return Optional.ofNullable(getOrNull(variableScope.getVariable(variableName)))
  }

  override fun set(value: Set<T>, isTransient: Boolean) {
    variableScope.setVariable(variableName, getTypedValue(value, isTransient))
  }

  override fun getLocalOptional(): Optional<Set<T>> {
    return Optional.ofNullable(getOrNull(variableScope.getVariableLocal(variableName)))
  }

  override fun setLocal(value: Set<T>, isTransient: Boolean) {
    variableScope.setVariableLocal(variableName, getTypedValue(value, isTransient))
  }

  override fun remove() {
    variableScope.removeVariable(variableName)
  }

  override fun removeLocal() {
    variableScope.removeVariableLocal(variableName)
  }
}
